### PR TITLE
Autostart the UwbConnector when the first event callback is registered

### PR DESCRIPTION
**Pull request by forwardpointer**
### Type

- [ ] Bug fix
- [X] Feature addition
- [ ] Feature update
- [ ] Breaking change
- [ ] Non-functional change
- [ ] Documentation
- [X] Infrastructure

### Goals

Autostart the UwbConnector when the first event callback is registered

### Technical Details

* There is a shared_mutex for reading/writing the callbacks weak_ptrs
* There is also a mutex and condition_variable for detecting the presence of callbacks

### Test Results

Verified that the basic ranging with random data scenario works with the simulator driver 

### Future Work

This design might need to be changed once we implement deregistering callbacks. 

### Checklist

- [ ] Build target `all` compiles cleanly.
- [ ] clang-format and clang-tidy deltas produced no new output.
- [ ] Newly added functions include doxygen-style comment block.

### CODE REVIEW

**Comment by forwardpointer** [comment]
I verified that the basic ranging scenario works with the simulator driver, but it's important to note that on the develop branch when I ran the same scenario the uwbsim.exe is able to exit by itself but not with this PR. This may be a bug

**Comment by forwardpointer** [comment]
> I verified that the basic ranging scenario works with the simulator driver, but it's important to note that on the develop branch when I ran the same scenario the uwbsim.exe is able to exit by itself but not with this PR. This may be a bug

Ok I verified that this is caused by an unused UwbConnector member in the UwbSimulator, so I removed it and uwbsim.exe exits properly now

File: windows/devices/uwb/UwbDeviceConnector.cxx [hunk 1]
```diff
@@ -731,19 +739,36 @@ UwbConnector::NotificationListenerStop()
 ::uwb::RegisteredCallbackToken*
 UwbConnector::RegisterDeviceEventCallbacks(std::weak_ptr<::uwb::UwbRegisteredDeviceEventCallbacks> callbacks)
 {
-    m_deviceEventCallbacks = callbacks;
+    {
+        std::lock_guard presenceLock{ m_callbacksPresentConditionVariableGate };
```
Comment (corbin-phipps) [inline windows/devices/uwb/UwbDeviceConnector.cxx:743]: Consider using `std::scoped_lock` instead of `std::lock_guard`, as `std::scoped_lock` is the newer and superior locking mechanism

File: windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx [hunk 2]
```diff
@@ -192,8 +195,23 @@ private:
...
+     * @brief Internal function to check if there are callbacks present
+     *
+     * @return true
+     * @return false
+     */
+    bool
+    CallbacksPresent();
+
 private:
+    mutable std::shared_mutex m_eventCallbacksGate;
```
Comment (corbin-phipps) [inline windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx:208]: Why is the `mutable` keyword needed?

File: windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx [hunk 3]
```diff
@@ -192,8 +195,23 @@ private:
...
+
+    mutable std::mutex m_callbacksPresentConditionVariableGate;
+    mutable std::condition_variable m_callbacksPresentConditionVariable;
+
     std::unordered_map<uint32_t, std::weak_ptr<::uwb::UwbRegisteredSessionEventCallbacks>> m_sessionEventCallbacks;
+
```
Comment (corbin-phipps) [inline windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx:214]: nit: Looks like this newline was added, not sure if intentional

**Review by corbin-phipps — CHANGES_REQUESTED** [review]

File: windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx (hunk 2 above)
Comment (abeltrano) [reply windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx:208]: This is typically done such that non-const locking functions can be used in `const` functions. It's technically cheating since the lock state does change, but the idea of `const` is more meant to reflect actual object state and not pseudo-state used to control concurrency.

File: windows/devices/uwb/UwbDeviceConnector.cxx [hunk 4]
```diff
@@ -29,7 +29,9 @@ class RegisteredCallbackToken
 
 UwbConnector::UwbConnector(std::string deviceName) :
     m_deviceName(std::move(deviceName))
-{}
+{
+    NotificationListenerStart();
```
Comment (abeltrano) [inline windows/devices/uwb/UwbDeviceConnector.cxx:33]: Why is this needed here? It seemed like the callback registration code would take care of calling this when necessary.

File: windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx [hunk 5]
```diff
@@ -192,8 +195,23 @@ private:
...
+     * @return true
+     * @return false
+     */
+    bool
+    CallbacksPresent();
+
 private:
+    mutable std::shared_mutex m_eventCallbacksGate;
+
+    mutable std::mutex m_callbacksPresentConditionVariableGate;
```
Comment (abeltrano) [inline windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx:210]: 2 separate mutexes aren't necessary since they're both protecting the same state. You can use the `std::shared_mutex` here as long as you change `m_callbacksPresentConditionVariable` to be a `std::condition_variable_any` and ensure that the lock you pass to `wait()` is a `std::unique_lock` of the `m_eventCallbacksGate`. When `wait()` is called, the mutex is atomically _released_, making it available to synchronize access to `m_sessionEventCallbacks` and `m_deviceEventCallbacks` in whatever way you want (ie. either with `std::shared_lock`, `std::unique_lock`, `std::scoped_lock`, etc.)

File: windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx (hunk 3 above)
Comment (abeltrano) [reply windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx:214]: I personally like grouping the concurrency related members together.

File: windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx [hunk 6]
```diff
@@ -150,6 +138,21 @@ public:
     SetApplicationConfigurationParameters(uint32_t sessionId, std::vector<::uwb::protocol::fira::UwbApplicationConfigurationParameter> applicationConfigurationParameters) override;
 
 private:
+    /**
+     * @brief Start listening for notifications.
+     *
+     * @return true If listening for notifications started successfully.
+     * @return false If listening for notifications could not be started.
+     */
+    bool
```
Comment (abeltrano) [inline windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx:147]: It made sense to have a `bool` return value when this was exposed publicly, but now that it's internal, can we just have this return `void` instead?

**Review by abeltrano — CHANGES_REQUESTED** [review]

File: windows/devices/uwb/UwbDeviceConnector.cxx [hunk 7]
```diff
@@ -480,6 +482,11 @@ UwbConnector::SetApplicationConfigurationParameters(uint32_t sessionId, std::vec
 void
 UwbConnector::HandleNotifications(std::stop_token stopToken)
 {
+    std::unique_lock lock{ m_callbacksPresentConditionVariableGate };
+    m_callbacksPresentConditionVariable.wait(lock, [this]() {
+        return CallbacksPresent();
+    });
```
[the remark this answers is not in the document]
Comment (forwardpointer) [reply windows/devices/uwb/UwbDeviceConnector.cxx:488]: gotcha, will add

File: windows/devices/uwb/UwbDeviceConnector.cxx (hunk 4 above)
Comment (forwardpointer) [reply windows/devices/uwb/UwbDeviceConnector.cxx:33]: This function is actually not called by anything else, as its only role is to start the wait process so I think the constructor will have to call it. Another way, is actually to get rid of the condition variable and just have the callback registration codes grab the mutex, check if the were callbacks registered before, and if there were none then also start the notification listener. I think this method actually makes a lot of sense

File: windows/devices/uwb/UwbDeviceConnector.cxx (hunk 1 above)
Comment (forwardpointer) [reply windows/devices/uwb/UwbDeviceConnector.cxx:743]: If I understand correctly then scoped_lock behaves the same as lock_guard but also has support for grabbing multiple mutexes in succession in the same line. Since we're also considering combining the two mutexes, I think I'll stick with lock_guard. (Maybe there's less overhead)

File: windows/devices/uwb/UwbDeviceConnector.cxx (hunk 4 above)
Comment (forwardpointer) [reply windows/devices/uwb/UwbDeviceConnector.cxx:33]: #203

File: windows/devices/uwb/UwbDeviceConnector.cxx (hunk 7 above)
[the remark this answers is not in the document]
Comment (forwardpointer) [reply windows/devices/uwb/UwbDeviceConnector.cxx]: After merging with #203 this is outdated, unless we want to revert back to using the condition variable

File: windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx [hunk 8]
```diff
@@ -2,10 +2,13 @@
 #ifndef UWB_DEVICE_CONNECTOR_HXX
 #define UWB_DEVICE_CONNECTOR_HXX
 
+#include <condition_variable>
 #include <cstdint>
 #include <future>
 #include <memory>
+#include <mutex>
```
Comment (abeltrano) [inline windows/devices/uwb/include/windows/devices/uwb/UwbDeviceConnector.hxx:9]: Is this still needed?

File: windows/devices/uwb/UwbDeviceConnector.cxx (hunk 4 above)
Comment (abeltrano) [inline windows/devices/uwb/UwbDeviceConnector.cxx:33]: Still doesn't seem like this is needed here. If it is, can you help me understand why?

**Review by corbin-phipps — DISMISSED** [review]

File: windows/devices/uwb/UwbDeviceConnector.cxx (hunk 4 above)
Comment (forwardpointer) [reply windows/devices/uwb/UwbDeviceConnector.cxx]: Missed this, thanks

**Review by abeltrano — APPROVED** [review]

**Review by corbin-phipps — APPROVED** [review]
